File: backend/src/main/java/nextpos/app/nextpos/service/interf/BarcodeScannerService.java
package nextpos.app.nextpos.service.interf;

import nextpos.app.nextpos.model.dto.request.ScannerRegistrationRequest;
import nextpos.app.nextpos.model.dto.request.BarcodeScanRequest;
import nextpos.app.nextpos.model.dto.response.BarcodeScanResponse;
import nextpos.app.nextpos.model.dto.response.ScannerRegistrationResponse;
import nextpos.app.nextpos.model.entity.BarcodeScanner;

import java.util.List;

public interface BarcodeScannerService {

    ScannerRegistrationResponse registerScanner(ScannerRegistrationRequest request);

    BarcodeScanResponse processBarcodeScan(BarcodeScanRequest request);

    void updateScannerStatus(String scannerId, String status);

    List<BarcodeScanner> getScannersByWarehouse(Long warehouseId);

    void disconnectScanner(String scannerId);

    BarcodeScanResponse validateAndProcessScan(String scannerId, String barcode);
}

File: frontend/src/services/barcodeScannerService.ts
import api from "./api";
import { useUserStore } from "@/stores/userStore";
import type {
  ScannerRegistrationRequest,
  ScannerRegistrationResponse,
  BarcodeScanRequest,
  BarcodeScanResponse,
  BarcodeScanner,
} from "@/types/barcodeScanner";

// Helper: build headers
const getHeaders = () => {
  try {
    const userStore = useUserStore();
    const companyId = userStore.currentUser?.companyId;
    const userId = userStore.currentUser?.id;

    if (!companyId) {
      throw new Error("Company information is missing.");
    }

    const headers: Record<string, string> = {
      "X-Company-Id": companyId.toString(),
    };

    if (userId) {
      headers["X-User-Id"] = userId.toString();
    }

    return headers;
  } catch (error) {
    console.error("Error getting user store:", error);
    throw new Error(
      "Unable to access user information. Please ensure you're logged in."
    );
  }
};

/**
 * Register a new barcode scanner
 */
export const registerScanner = (data: ScannerRegistrationRequest) => {
  const headers = getHeaders();
  return api.post<ScannerRegistrationResponse>("/scanner/register", data, {
    headers,
  });
};

/**
 * Process barcode scan via REST API
 */
export const processBarcodeScan = (data: BarcodeScanRequest) => {
  const headers = getHeaders();
  return api.post<BarcodeScanResponse>("/scanner/scan", data, {
    headers,
  });
};

/**
 * Get all scanners for a warehouse
 */
export const getWarehouseScanners = (warehouseId: number) => {
  const headers = getHeaders();
  return api.get<BarcodeScanner[]>(`/scanner/warehouse/${warehouseId}`, {
    headers,
  });
};

/**
 * Disconnect a scanner
 */
export const disconnectScanner = (scannerId: string) => {
  const headers = getHeaders();
  return api.post<void>(
    `/scanner/${scannerId}/disconnect`,
    {},
    {
      headers,
    }
  );
};

/**
 * Get scanner by ID
 */
export const getScannerById = (scannerId: string) => {
  const headers = getHeaders();
  return api.get<BarcodeScanner>(`/scanner/${scannerId}`, {
    headers,
  });
};
